wallet/wallet.go

```diff
@@ -322,379 +322,414 @@ func (w *Wallet) activeData(dbtx walletdb.ReadTx) ([]btcutil.Address, []wtxmgr.C
+	var birthdayStamp *waddrmgr.BlockStamp
 	birthday := w.Manager.Birthday()
 
-	// TODO(jrick): How should this handle a synced height earlier than
-	// the chain server best block?
+	// We'll start by getting our chain client and database transaction
+	// ready.
+	chainClient, err := w.requireChainClient()
+	if err != nil {
+		return nil, err
+	}
+	tx, err := w.db.BeginReadWriteTx()
+	if err != nil {
+		return nil, err
+	}
+	ns := tx.ReadWriteBucket(waddrmgrNamespaceKey)
 
-	// When no addresses have been generated for the wallet, the rescan can
-	// be skipped.
-	//
-	// TODO: This is only correct because activeData above returns all
-	// addresses ever created, including those that don't need to be watched
-	// anymore.  This code should be updated when this assumption is no
-	// longer true, but worst case would result in an unnecessary rescan.
-	if isInitialSync || isRecovery {
-		// Find the latest checkpoint's height. This lets us catch up to
-		// at least that checkpoint, since we're synchronizing from
-		// scratch, and lets us avoid a bunch of costly DB transactions
-		// in the case when we're using BDB for the walletdb backend and
-		// Neutrino for the chain.Interface backend, and the chain
-		// backend starts synchronizing at the same time as the wallet.
-		_, bestHeight, err := chainClient.GetBestBlock()
-		if err != nil {
-			return err
-		}
+	// Determine the latest height known to the chain backend. This will
+	// serve as our finishing point when attempting to find our birthday
+	// stamp.
+	var latestCheckptHeight int32
+	if len(w.chainParams.Checkpoints) > 0 {
+		latestCheckptHeight = w.chainParams.
+			Checkpoints[len(w.chainParams.Checkpoints)-1].Height
+	}
+	_, bestHeight, err := chainClient.GetBestBlock()
+	if err != nil {
+		return nil, err
+	}
 
-		checkHeight := bestHeight
-		if len(w.chainParams.Checkpoints) > 0 {
-			checkHeight = w.chainParams.Checkpoints[len(
-				w.chainParams.Checkpoints)-1].Height
-		}
+	logHeight := latestCheckptHeight
+	if bestHeight > logHeight {
+		logHeight = bestHeight
+	}
 
-		logHeight := checkHeight
-		if bestHeight > logHeight {
-			logHeight = bestHeight
+	log.Infof("Catching up block hashes to height %d, this will take a "+
+		"while...", logHeight)
+
+	// isCurrent is a helper function that we'll use to determine if the
+	// chain backend is currently synced.
+	isCurrent := func(bestHeight int32) bool {
+		switch c := chainClient.(type) {
+		case *chain.NeutrinoClient:
+			return c.CS.IsCurrent()
 		}
+		return bestHeight >= latestCheckptHeight
+	}
 
-		log.Infof("Catching up block hashes to height %d, this will "+
-			"take a while...", logHeight)
+	for height := w.Manager.SyncedTo().Height; birthdayStamp == nil; height++ {
+		// If we've reached our best height and we're not current, we'll
+		// wait for blocks at tip to ensure we find a valid birthday
+		// stamp.
+		for height == bestHeight && !isCurrent(bestHeight) {
```

Comment:
@wpaulino wanted to point out that I am not sure the "!isCurrent(bestHeight)" is needed in the condition. This is because at this point the birthday stamp is null and we are certain that we are not "current".

Reply:
It's not meant to check if we're current, but if the chain backend is current or not. This is how it works in the current diff at least, since it may be the case that the chain backend is still syncing while `lnd` is connected to it.

Reply:
> It's not meant to check if we're current, but if the chain backend is current or not

This is why I was suggesting to remove this condition. as long as "height == bestHeight" we can assume the chain backend is not current since at this point of the code the birthday stamp is still null. Also think it is better to move this conditional waiting block to the end of the for loop. First process the block then wait.